PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,317 @@
+
+        self.nonConsumablePurchases = Set(subscriberData.nonSubscriptions.keys)
+        self.nonSubscriptionTransactions = subscriberData.nonSubscriptionTransactions
+
+        self.entitlements = EntitlementInfos.init(entitlementsData: subscriberData.entitlements, purchasesData: subscriberData.allPurchases, dateFormatter: DateFormatter.dateFormatter, requestDate: requestDate)
+
+        self.expirationDatesByProduct = subscriberData.expirationDatesByProduct
+        self.purchaseDatesByProduct = subscriberData.purchaseDatesByProduct
+    }
+
+    // TODO after migration make this internal
+    @objc public class func currentSchemaVersion() -> String { "2" }
+
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? { expirationDatesByProduct[forProductIdentifier] ?? nil }
+
+     /**
+     Get the latest purchase or renewal date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for subscription product
+    
+     @return The purchase date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func purchaseDate(forProductIdentifier: String) -> Date? { purchaseDatesByProduct[forProductIdentifier] ?? nil }
+
+    /**
+     Get the expiration date for a given entitlement.
+    
+     @param forEntitlement The id of the entitlement.
+    
+     @return The expiration date for the passed in `entitlement`, can be `nil`
+     */
+    @objc public func expirationDate(forEntitlement: String) -> Date? { entitlements[forEntitlement]?.expirationDate }
+
+    /**
+     Get the latest purchase or renewal date for a given entitlement identifier.
+    
+     @param forEntitlement Entitlement identifier for entitlement
+    
+     @return The purchase date for `entitlement`, `nil` if product never purchased
+     */
+    @objc public func purchaseDate(forEntitlement: String) -> Date? { entitlements[forEntitlement]?.latestPurchaseDate }
+
+    // TODO after migration make this internal and remove objc rename
+    @objc(JSONObject) public func jsonObject() -> [AnyHashable: Any] {
+        return originalData.merging(["schema_version": PurchaserInfo.currentSchemaVersion()], uniquingKeysWith: { (current, _) in current })
+    }
+
+    public override func isEqual(_ object: Any?) -> Bool {
+        guard let other = object as? PurchaserInfo else {
+            return false
+        }
+
+        var selfJson = self.jsonObject()
+        selfJson.removeValue(forKey: "request_date")
+        var otherJson = other.jsonObject()
+        otherJson.removeValue(forKey: "request_date")
+
+        return NSDictionary(dictionary: selfJson).isEqual(to: otherJson)
```

Comment:
also, do we need these to be NSDictionaries?
would doing something `selfJson == otherJson` not work? is it because of the `AnyHashable` thing?

Reply:
> also, do we need these to be NSDictionaries?
> would doing something `selfJson == otherJson` not work? is it because of the `AnyHashable` thing?

yes, because the keys are `Any` which doesn't conform to equatable.


> should this be `NSDictionary(dictionary: selfJson).isEqual(to: NSDictionary(dictionary: otherJson))`?

that version of isEqual takes [AnyHashable: Any], but making a dictionary feels safer/clearer to me so i'll update it anyway